### PR TITLE
[SPARK-48143][SQL] Use lightweight exceptions for control-flow between UnivocityParser and FailureSafeParser

### PR DESCRIPTION
**Pull request by vladimirg-db**
# What changes were proposed in this pull request?
New lightweight exception for control-flow between UnivocityParser and FailureSafeParser to speed-up malformed CSV parsing.

This is a different way to implement these reverted changes: https://github.com/apache/spark/pull/46478

The previous implementation was more invasive - removing `cause` from `BadRecordException` could break upper code, which unwraps errors and checks the types of the causes. This implementation only touches `FailureSafeParser` and `UnivocityParser` since in the codebase they are always used together, unlike `JacksonParser` and `StaxXmlParser`. Removing stacktrace from `BadRecordException` is safe, since the cause itself has an adequate stacktrace (except pure control-flow cases).

### Why are the changes needed?
Parsing in `PermissiveMode` is slow due to heavy exception construction (stacktrace filling + string template substitution in `SparkRuntimeException`)

### Does this PR introduce _any_ user-facing change?
No, since `FailureSafeParser` unwraps `BadRecordException` and correctly rethrows user-facing exceptions in `FailFastMode`

### How was this patch tested?
- `testOnly org.apache.spark.sql.catalyst.csv.UnivocityParserSuite`
- Manually run csv benchmark
- Manually checked correct and malformed csv in sherk-shell (org.apache.spark.SparkException is thrown with the stacktrace)

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by vladimirg-db** [comment]
@HyukjinKwon @cloud-fan @dongjoon-hyun hi! New implementation for the reverted https://github.com/apache/spark/pull/46478

**Comment by vladimirg-db** [comment]
Checked the benchmark, the perf is the same as it was on the reverted version

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 1]
```diff
@@ -316,7 +316,7 @@ class UnivocityParser(
       throw BadRecordException(
         () => getCurrentInput,
         () => Array.empty,
-        QueryExecutionErrors.malformedCSVRecordError(""))
+        MalformedCSVRecordException(""))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:319]: This approach is like using a placeholder exception first and then replacing it with the actual error at a later phase. The current implementation needs a dedicate placeholder exception for each real error.

Shall we make it more general? We can have a special wrapper exception, which does not have the cause field and can lazily produce the actual error? Just like the updated `BadRecordException` in the previous attempt.

Comment (vladimirg-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:319]: Do you mean, to use some other exception _in place_ of `BadRecordException` between the `UnivocityParser` and `FailureSafeParser`? Looks like a cleaner idea! But it seems that in that case we're gonna end up with two similar mechanisms for the same thing. And eventual refactoring for this issue does not seem easy, considering how extensively the parsers and `BadRecordException` are used in the codebase...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala [hunk 2]
```diff
@@ -67,16 +67,30 @@ case class PartialResultArrayException(
...
+ * @param cause the actual exception about why the record is bad and can't be parsed. It's better
+ *                      to use `LazyBadRecordCauseWrapper` here to delay heavy cause construction
+ *                      until it's needed.
  */
 case class BadRecordException(
     @transient record: () => UTF8String,
     @transient partialResults: () => Array[InternalRow] = () => Array.empty[InternalRow],
-    cause: Throwable) extends Exception(cause)
+    cause: Throwable) extends Exception(cause) {
+  override def getStackTrace(): Array[StackTraceElement] = new Array[StackTraceElement](0)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala:82]: If `BadRecordException` can be a user-facing exception, doesn't this change degrade the error message?

Comment (vladimirg-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala:84]: If the cause is other than `LazyBadRecordExceptionCauseWrapper`, the stacktrace from the cause will be enough.

Comment (vladimirg-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala:84]: Added a doc for that

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**